design/secrets_provider_milestone1_2_design.md

```diff
@@ -0,0 +1,367 @@
+
+### Motivation
+
+The current implementation makes the deployment of the Secrets Provider does not support secret rotation and makes the deployment of the Secrets Provider deeply coupled to the customer's applications which not all customers want.
+
+### Requirement
+
+The solution should support the following capabilities:
+
+1. Deployment, deploy as separate Conjur entity; serve K8s Secrets to many Applications
+2. Rotation, update K8s Secret with Conjur secret values (on trigger or on a timely, configurable basis)
+3. Support removal of deployment
+
+## Solution
+This solution will be divided into milestones.
+
+- Milestone 1: Deployment, deploy as as separate entity
+- Milestone 2: Handle rotation, updating K8s Secrets with Conjur secret value (configurable amount of time or on a trigger)
+
+
+
+### Milestone 1: Deployment
+
+#### Design
+
+The Secrets Provider for K8s will no longer run as an *init* container attached to the application pod, rather run as a separate single Job in a namespace. A Job creates a Pod to perform a task and terminates upon completion. Once the Pod is spun up, it will authenticate to Conjur/DAP in the same way it does today, via authn-k8s. The Pod will then fetch the K8s secret denoted with a label and update them with the Conjur secrets they require.
+
+|      | Deployment                  | Elaboration                                                  | Pros                                 | Cons                                                         |
+| ---- | --------------------------- | ------------------------------------------------------------ | ------------------------------------ | ------------------------------------------------------------ |
+| 1    | Run as Job                  | Run as a K8s Job that deploys a Pod that terminates after task completion | - Doesn't waste customer resources   | - Requires addition of "Job" to application identity granularities. Otherwise a Job will limit us to only using Namespace / Service Account<br /><br />- Deployment will need to be changed (from Job → Pod) for Milestone 2 (support secrets rotation) |
+| 2    | Run as non-terminating  Pod | Run as a Pod that does not terminate automatically after task completion | - Simple<br />- Minimal code changes | - Wastes resources because we aren't updating K8s secrets and up all the time<br /><br />- For **deployment**, **deploymentConfig** and **statefulSet** segregation of duties, this solution is not interchangeable with our other secrets retrieval solutions without policy changes. (e.g. secrets provider as init-container, conjur authn k8s secrets, any of our clients or Conjur REST API) |
+
+**Decision:** Run as a Job
```

Comment:
Isn't this limiting our ability to be extensible? Remember we still want to keep good engineering skills such as SOLID, KISS etc.

Reply:
i would like to challenge this decision. it seems that for milestone 2 we will go back to not using a Job? if this is the case then why invest in this only for the first milestone? are we planning to ship this milestone as well? 

Also - why are you referring to the other option as a `non-terminating pod`? why is it not terminating? You can have a pod in a Deployment(Config) that runs with `CONTAINER_MODE=init` as an env var and it will exit the pod once the secrets retrieval is done. This option is also extensible to the next milestone (where you will just change the value of `CONTAINER_MODE`.

Reply:
btw, if you do go with the above i would change the value of `CONTAINER_MODE` to something else (as it is not an init container) and change our code to exit the container also with that value.